### PR TITLE
Data (pre-)generation when scaling both model and data

### PR DESCRIPTION
**Pull request by fluegelk**
Add functionality to (pre-)generate balanced datasets where the training set is split across a fixed number of nodes.
These datasets can be written to and read from HDF5 files to avoid the memory overhead when generating large global datasets.

### Data Generation
Call `python specialcouscous/scaling_dataset.py` with the same CLI arguments as the training scripts but additionally specify:
- `--n_train_splits` the number of local training sets to split into (usually: the largest number of nodes this dataset shall be used with)
- `--data_root_path` the root path to store generated datasets at (default `./data`)
- pass `--override_data` to override should there already be a dataset of the same name

### Load Dataset from File
Call `scaling_dataset.load_and_verify_dataset(args, rank)` with the same CLI arguments `args` as the dataset was generated for to get local train set for rank `rank` and global test set.

### CODE REVIEW

**Review by mcw92 — APPROVED** [review]
I went through the code and made a couple of very minor comments mainly concerning the style of docstrings + comments in the code (for consistency throughout the codebase) ☺️. 
Regarding your questions and remarks from yesterday:

> Allerdings sollten wir uns nochmal Gedanken machen, welche Parameter wir verwenden wollen, konkret wie viele lokale Train-Samples pro Knoten und wie viele globale Test-Samples es geben soll. Da die Test-Samples anteilig vom globalen Dataset gesplittet werden skaliert auch das Test-Set mit der maximalen #Ranks. -> Wollen wir das?

Alles gute Fragen, meine ursprüngliche Idee war im Grunde, dass jeder Knoten ein lokal unterschiedliches n6m4 / n7m3 Train Set und ein global einheitliches Test Set bekommt, wobei ich ursprünglicherweise einfach naiv dieselbe Anzahl an Samples im globalen Test Set wie bei Weak Scaling genommen hätte, also ohne das Test Set mit der Anzahl an maximalen Ranks zu skalieren. Aber wenn das mit dem Speicher sowieso in der Form nicht hinhaut, sind die Überlegungen im Grunde hinfällig.

> Selbst mit Vorgenerieren des Datasets und Extra-Large Memory Nodes können wir die Weak-Scaling Datasets (n6m4 & n7m3) leider nicht auf 64 Knoten hochscalieren. Die größten Knoten haben 4096 GB, die reinen Feature-Daten bräuchten aber 5120 GB (64 (Knoten) * 10^6 (Samples pro Knoten) * 10^4 (Features) * 64bit (float64)). Mit den Targets und Overhead durch das Partitionieren brauchen wir sogar noch mehr (hier könnte man ggf. noch optimieren). Mir fallen folgende Optionen ein:
> 1. `max_ranks` reduzieren und z.B. nur bis 16 oder 32 Knoten skalieren
> 2. kleineres Dataset pro Knoten verwenden (evtl wie bei Chunking mit 64 Knoten)
> 3. versuchen `make_classification` von `float64` auf `float32` zu ändern, allerdings scheint es dafür keine bestehende Option zu geben, man müsste also vermutlich den `sklearn`-Code direkt ändern. Ich vermute das lohnt sich eher nicht.

Über Option 3 hatte ich auch schon mal nachgedacht, als ich Speicherprobleme beim Re-Evaluieren der Checkpoints hatte, bin aber auch zu dem Entschluss gekommen, dass sich das vorerst und vermutlich auch langfristig nicht wirklich lohnt. 
So spontan kann ich nicht wirklich sagen, ob ich Option 1 oder 2 besser finde. Generell würde ich sagen, wir sollten so groß gehen wie wir können, also auf die 64 Knoten und dann den Datensatz so groß machen, wie es der Speicher zulässt. Ich bin mir sowieso nicht so ganz sicher, inwieweit die Vergleichbarkeit mit den Weak Scaling Experimenten gegeben ist bzw. sein muss. Vermutlich geht es einfach generell mehr um irgendwelche Trends, die wir ggf. beobachten, wenn man Daten + Modell gleichzeitig skaliert. In dem Fall wäre es auch egal, wenn die Test Set Größe mit der maximalen Anzahl an Ranks skaliert, solange es innerhalb des Experiments konsistent ist.
Möglicherweise haben wir auch genug Compute, um sowohl Option 2 als auch 3 auszuprobieren.

Was denkst du darüber?

Bzgl. des To Dos im Test: Ich glaube nicht, dass wir aktuell unbedingt Zeit darauf verwenden müssen, uns einen besseren Check zu überlegen, das sollte ja im Prinzip passen (ich weiß, das ist eigentlich kein Grund 🙈). Vielleicht einfach ein Issue dazu anlegen, damit es erstmal irgendwo noch dokumentiert ist?

File: specialcouscous/scaling_dataset.py [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+    Parameters
+    ----------
+    global_train_set : SyntheticDataset
+        The global train set (used only to extract root attributes)
+    local_train_sets : dict[int, SyntheticDataset]
+        pass
+    global_test_set : SyntheticDataset
+        pass
+    additional_global_attrs : dict[str, Any]
+        pass
```
[the remark this answers is not in the document]
Comment (fluegelk) [reply specialcouscous/scaling_dataset.py:167]: Good catch, these are still missing
